### ISSUE TITLE
NumPy 1.21 requires PyPy 7.3.4 or later

### ISSUE THREAD

**Issue by filipinascimento** [numpy/numpy#19383]
Numpy giving errors on import inside a build environment (github action) for a specific python version/platform.

### Reproducing code example:
After pip install numpy:

```python
import numpy as np
# that is it
```

### Error message:
Should start reading from `File "/tmp/pip-req-build-boe77mfe/setup.py", line 33, in __str__` 

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/tmp/pip-req-build-boe77mfe/setup.py", line 109, in <module>
    extra_link_args=extraLinkOptions,
  File "/opt/pypy/pypy3.7-7.3.3/site-packages/setuptools/__init__.py", line 144, in setup
    return distutils.core.setup(**attrs)
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/opt/pypy/pypy3.7-7.3.3/site-packages/wheel/bdist_wheel.py", line 223, in run
    self.run_command('build')
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/command/build.py", line 135, in run
    self.run_command(cmd_name)
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/opt/pypy/pypy3.7-7.3.3/site-packages/setuptools/command/build_ext.py", line 87, in run
    _build_ext.run(self)
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/command/build_ext.py", line 343, in run
    self.build_extensions()
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/command/build_ext.py", line 452, in build_extensions
    self._build_extensions_serial()
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/command/build_ext.py", line 477, in _build_extensions_serial
    self.build_extension(ext)
  File "/opt/pypy/pypy3.7-7.3.3/site-packages/setuptools/command/build_ext.py", line 208, in build_extension
    _build_ext.build_extension(self, ext)
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/command/build_ext.py", line 537, in build_extension
    depends=ext.depends)
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/ccompiler.py", line 566, in compile
    depends, extra_postargs)
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/ccompiler.py", line 341, in _setup_compile
    pp_opts = gen_preprocess_options(macros, incdirs)
  File "/opt/pypy/pypy3.7-7.3.3/lib-python/3/distutils/ccompiler.py", line 1076, in gen_preprocess_options
    pp_opts.append("-I%s" % dir)
  File "/tmp/pip-req-build-boe77mfe/setup.py", line 33, in __str__ 
    import numpy
  File "/tmp/pip-req-build-boe77mfe/.eggs/numpy-1.21.0-py3.7-linux-x86_64.egg/numpy/__init__.py", line 160, in <modu
    from . import polynomial
  File "/tmp/pip-req-build-boe77mfe/.eggs/numpy-1.21.0-py3.7-linux-x86_64.egg/numpy/polynomial/__init__.py", line 11
    from .polynomial import Polynomial
  File "/tmp/pip-req-build-boe77mfe/.eggs/numpy-1.21.0-py3.7-linux-x86_64.egg/numpy/polynomial/polynomial.py", line 
    polydomain = np.array([-1, 1])
TypeError: array() missing required argument 'object' (pos 0)
```

### NumPy/Python version information:
This is running in a github action on a manylinux build environment, so not so easy to run this specific command. However 
sys.version is:
```
3.7.9 (7e6e2bb30ac5, Nov 18 2020, 10:55:52)
[PyPy 7.3.3-beta0 with GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
```

**Comment by seberg** [comment]
Hmm, I wonder if we can document this.  PyPy 7.3.3 did not have support for a Python 3.7 C-API feature that NumPy now uses.  If you update to 7.3.5 I am very sure that will work.  7.3.4 may also be good enough (but I am not sure).

**Comment by filipinascimento** [comment]
Thanks for the explanation. Currently we are dropping pypy support for our package, but I will try pypy 7.3.5 later. Unfortunately this is the version available in the cibuildwheel build environment for github actions, so I may need to check there how to update pypy on the building environment. thanks!
It would be great if someone could cross reference other problems related to pypy versions < 7.3.5). Not sure if this is documented anywhere.

**Comment by seberg** [comment]
Ping @mattip, just in case you have any tip about pypy CI/buildwheel setup.

**Comment by mattip** [comment]
> Unfortunately this is the version available in the cibuildwheel build environment for github actions,

Are you sure? At least according to [this PR](https://github.com/pypa/cibuildwheel/pull/671) pypy7.3.5 should be supported.

**Comment by joerick** [comment]
The linked PR is only available in v2.0.0, which is currently in prerelease. The only known issues with the current alpha relate to PyPA/build, which is a new feature you're probably not using, so I think you could use it if you want to!

**Comment by seberg** [comment]
Going to close this, I assume getting a new enough PyPy version for NumPy is not an issue anymore (and it is not a NumPy issue per-se).
